### PR TITLE
Add Webhook to Form

### PR DESCRIPTION
**Pull request by Soare-Robert-Daniel**
Closes #.

### Summary

Add Webhook triggers to Form.

### Screenshots 

image

image

image

----

### Workflow Explanation

The idea of the webhook is that after a form has been submitted, it will ping a given URL which will trigger an external action/workflow (e.g.: send an SMS, email, a Slack message).

In this feature, we will support Zappier and IFTTT. The way they function is the same regarding the API call. (which means we do not need separate codes for them).

For the features to work, we need three parts:
- the URL of the webhook
- the method: GET, POST,  PATCH
- the headers.

Those settings will be managed in the Webhook editor available in Form at Form Options > Webhook (check screenshots from above).

When a user submits a form, the data of the form will be prepared in a payload with the structure (in JSON format): `key: value`. The `value` will be the field values (⚠️ for File Input the values will be the URI of the saved file (given always as an `array`) - if they are saved on Media Library the URL will be provided. Otherwise, the location on the disk will be given). 

The `key` should be confused with the label. By default, the key will a unique id that resembles the following structure `input-1234234` (this is not enforced). The key is generated using the field unique internal id.

ℹ️ Each field now has a new option named `Mapped Name` which is used as a `key` instead of the default one if present.

#### Example with default keys:

image

#### Example with Mapped Name

image
 
ℹ️ The main role of Mapped Name is to offer a uniform way of reusing the webhook. Suppose you have three different forms with different types and numbers of fields. And a webhook that takes a phone number and sends an SMS. So for each form, you create the field for the user to introduce the phone number then you set the Mapped Name as `phone_number`, then in the webhook provider, you set a rule to extract the `phone_number` from the request.

#### In Zappier you have something like this
Screenshot 2023-06-30 at 16 14 26

⚠️ Both Zappier and IFTTT have some cumbersome documentation and website. I strongly suggested some tutorials before starting to document/test this feature.

### Zapier

Zapier is one of the biggest services for integrating things with many services. Workflows in Zapier are called Zaps. When you create a Zap, it shows a drag-and-drop editor in which you put different Apps to connect. To integrate with Form, you need to insert the `Catch Hook in Webhooks by Zapier` (this is a premium feature) it will be marked with a `Trigger` label.

When you press on it, on the right side, a sidebar will open, and the Test tab will appear, which has a section named `Your webhook URL` with a URL that you need to introduce in the Form Webhook feature (see pictures above) on URL field in Webhook Editor.

image

You will need the URL from Zapier to connect, and you are ready to go. When you submit a form, all its data will be sent to the webhook (you can check the pictures above).

Regarding the usage of the data that we send, it will be more on the user side on how to configure it. For this, let's put references for tutorials. 

To test this, you can attach a Gmail workflow named `Gmail` with `Create Draft` event in App & event. Then you can set up the values in the Action tab.

image

:information: After you send some submissions and return to the Zapier interface, you will notice that after you press on the field, it will show some suggestions about the values. Zapier parses the fields and offers autocompletion so it can be easier to set up the app.

### IFTTT

🔥 [Video Guide](https://github.com/Codeinwp/otter-internals/issues/96#issuecomment-1630333367)

IFTTT is a Zapier alternative but with an ugly desktop interface (personal opinion). The setup is the same with the differences that workflow is named Applets and the Webhook App is called  `Webhooks`. Unlike Zapier, getting the URL is more cumbersome. To get the URL, you need to go to [this page](https://ifttt.com/maker_webhooks) and press Documentation (you need to be logged in), and the page you are going to see it will contain all the details.

image

ℹ️  When you first insert that Webhooks in your Applet, it will ask you to introduce an `Event Name`; this name will be used to identify the webhook.

In the documentation, you will find an URL like this: `https://maker.ifttt.com/trigger/{event}/json/with/key/c1kinRqHROD2Ds...`, in which you replace the `{event}` with your webhook name. If you set the webhook name as `test_1`, the link will be: `https://maker.ifttt.com/trigger/test_1/json/with/key/c1kinRqHROD2Ds...`

You take this URL and put it in the URL field in the Webhook Editor of the Form Block. As you can see, is the same thing as Zapier.

ℹ️ Do not forget to add the Header: `Content-Type: application/json`

To this up, I recommend connecting your account with Google and adding the `Send yourself an email` via Gmail, which practically will send an email with data with JSON payload.

image

ℹ️  Even if IFTTT does the same thing as Zapier, it does not have the same features regarding the UX. In Zapier, things are in the same Editor, while in IFTTT, you will have multiple navigations to check things.

### Change the payload data with PHP Hooks

Using PHP, users can change the data of the payload that will be sent to the webhook.

The name of the filter is `themeisle_blocks_form_webhook_payload`. It takes the three parameters: the current content as `array`, the form data of type `Form_Data_Request` ([definition](https://github.com/Codeinwp/otter-blocks/blob/master/inc/integrations/api/form-request-data.php)), and the current used webhook data, which have `method, url, headers` as `array keys`.

Example:

```php

         /**
	 * Add extra data to the payload that marks the IFTTT webhook provider.
	 *
	 * @param mixed $payload The payload.
	 * @param Form_Data_Request $form_data The form data.
	 * @param mixed $webhook The webhook.
	 * @return mixed
	 */
	function mark_ifttt_provider( $payload, $form_data, $webhook ) {
		if ( str_contains( $webhook['url'], 'maker.ifttt.com' ) ) {
			$payload['webhook_provider'] = 'ifttt';
		}

		return $payload;
	}

        add_filter( 'otter_form_webhook_payload', 'mark_ifttt_provider', 10, 3 );
```

----

### Test instructions

1. Create a form with some fields.
2. Create a webhook (Zappier or IFTTT) and make the setup via Webhook Editor.
3. Test if the data has been received (check the webhook provider)
4. Go back to the fields and set a `Mapped Name` to set a special name for the field --- you can read more about this above.
5. Test again.

⚠️ The job of the Form is to call the webhook and trigger a workflow. The payload will be served like the description above. For this version, we do not aim to have special integration (like preparing the payload in a way to work with X service from Zappier).  

---- 

### Checklist before the final review

- [ ] Included E2E or unit tests for the changes in this PR.
- [x] Visual elements are not affected by independent changes.
- [x] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [x] It loads additional script in frontend only if it is required.
- [x] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [x] In case of deprecation, old blocks are safely migrated.
- [x] It is usable in Widgets and FSE.
- [x] Copy/Paste is working if the attributes are modified.
- [x] PR is following [the best practices]()

### CODE REVIEW

File: plugins/otter-pro/inc/plugins/class-form-pro-features.php [hunk 1]
```diff
@@ -347,6 +348,135 @@ public function send_autoresponder( $form_data ) {
 		}
 	}
 
+	/**
+	 * Send autoresponder email to the subscriber.
+	 *
+	 * @param Form_Data_Request|null $form_data The files to load.
+	 * @since 2.4
+	 */
+	public function trigger_webhook( $form_data ) {
```
Comment (HardeepAsrani) [inline plugins/otter-pro/inc/plugins/class-form-pro-features.php:357]: Just a reminder that we need to add a filter for the data being sent to webhook, with the webhook & form id being a parameter

**Comment by HardeepAsrani** [comment]
@Soare-Robert-Daniel I tested it with Zapier and IFTTT, it works nicely. Here are few things we can do:

- Remove the `!` from Webhook Saved notification, just a NIT
- Add the filter as we discussed in L10 so people can modify the data that is sent to the hook.
- Have a small testing doc ready for QA and something that Docs team can use to create docs around this feature:
  - How to connect and use with IFTTT
  - How to connect and use with Zapier
  - How to use filter to modify the payload
- And we can also add a guidelink to the feature Form sidebar so people can see how it can be used.

Let me know your thoughts.

File: plugins/otter-pro/inc/plugins/class-form-pro-features.php [hunk 2]
```diff
@@ -347,6 +349,150 @@ public function send_autoresponder( $form_data ) {
...
+
+				foreach ( $headers_pairs as $pair ) {
+					if ( empty( $pair['key'] ) || empty( $pair['value'] ) ) {
+						continue;
+					}
+
+					$headers[] = $pair['key'] . ': ' . $pair['value'];
+				}
+
+				$payload = apply_filters( 'themeisle_blocks_form_webhook_payload', array(), $form_data, $webhook );
```
Comment (HardeepAsrani) [inline plugins/otter-pro/inc/plugins/class-form-pro-features.php:401]: A NIT: we can keep the same naming for filters as we do for rest of the code: https://github.com/Codeinwp/otter-blocks/blob/ce56737db046cb5d2bfad6af8656dec8fe1641be/inc/server/class-form-server.php#L241

**Review by HardeepAsrani — CHANGES_REQUESTED** [review]
Just some small changes and then we can put it for QA and docs creation.

**Comment by rodica-andronache** [comment]
@Soare-Robert-Daniel I've tested both integrations (Zappier and IFTTT) and everything's working well 🚀

**Pull request merged**